### DISCUSSION TITLE
[Quiz] prototype 퀴즈

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#40] [프론트 기술 질문 공간]
```JSX
function 메롱() {
  console.log("메롱");
}
//
// A.
//
function App1() {
  this.render = () => {
    메롱();
  };
}
//
// B. 
//
function App2() {}

App2.prototype.render = () => {
  메롱();
};
//
// 클래스
//
class App3 {
  render() {
    메롱();
  };
}

const app1 = new App1();
app1.render(); // "메롱"
const app2 = new App2();
app2.render(); // "메롱"
const app3 = new App3();
app3.render(); // "메롱"
```
1. 위 코드 A. 와 B. 중 메소드 선언의 작동원리가 클래스의 그것과 같은 것은 무엇인가요?
2. [쉽지않음] 그렇다면 ES6 class 에서는 왜 해당 메소드 선언 방식을 디폴트로 설정했을까요? 프로젝트가 커지는 경우 잠정적으로 해당 코드들이 어떤 차이가 발생할 수 있는지 알려주세요.
> 힌트: App들의 인스턴스를 추가하는 경우 render 메소드는 어떻게 될까요?

도움이 되는 글: https://velog.io/@gtobio11/Javascript-Prototype-methods-vs-Object-methods

**Comment by michoball** [comment]
B가 아닐까 싶습니다. 

A와 같은 방법을 사용하면 new 키워드로 새로운 객체를 생성할 때 render라는 메소드를 복사해서 새롭게 메모리에 할당한다고 알고 있습니다. 때문에 새로 만든 객체에서 render 메소드를 변경하면 해당 변경내용은 그 객체안에만 반영되게 됩니다. 

프로젝트가 커지면서 계속 위와 같은 식으로 객체를 생성하면 불필요하게 메모리를 계속 할당해야하므로 비효율적이라서 그렇지 않나 싶습니다.

**Comment by WooGie911** [comment]
1.B라고 생각합니다.


코드를 살펴보면 
`
App2.prototype.render = () => {
  메롱();
};
`
부분에서 class App2 의 render()라는 내장함수를 재할당하여 app2.render() 시 메롱() 함수가 실행됩니다. 
class App3 에서 도 B와 마찬가지로 공통 메소드를 변화 시킴으로써 같은 효과를 내는 것으로 보입니다.


2.객체의 생성자 함수에서 메소드를 추가하게 되면 각 메소드들이 사본을 가지게 되어 시스템의 처리 시간과 메모리에 비효율적입니다. 그러나 객체의 prototype을 통해서 메소드를 추가하면 오직 단 하나의 버전의 메소드가 존재하게 되어 처리 시간과 메모리 모두 이점을 가질 수 있습니다.

**Comment by skyhanull** [comment]
1.B인 것 같습니다 A는 기본적으로 사용하는 함수형 같아서..?
2. 이부분은 잘모르겠습니다..ㅠㅠ

**Comment by BaekYuri** [comment]
1. A와 같다고 생각합니다. B와 같은 경우에는 프로토타입에 메소드가 선언되어있기 때문에 이 메소드를 변경한다면 기존에 생성된 객체들의 render함수가 다 같이 동일하게 변하지만 A와 클래스 같은 경우에는 프로토타입에 메소드가 선언되지 않고, 그냥 객체 내부에 생성되는 것으로 보입니다.
2. extends를 받는 부분이 프로토타입으로 들어가야 해서..? 라고 생각합니다. funtion Cat extends Animal 이라고 할 때, Animal 부분이 프로토타입으로 들어가고, Cat 부분이 객체 부분으로 들어가야 상속이 제대로 작동할 것이라고 생각합니다.

**Reply by javascriptbiggosoo** [reply]
(수정 전)유일한 정답자.. 화살표 함수로 작성하니 문제가 꼬였었네요 문제가 틀리고 이 답이 맞았습니다!

**Comment by javascriptbiggosoo** [comment]
1. B입니다.
2. `this`를 이용한 메소드 선언 방식은 constuctor 함수로 object를 만들때마다 똑같은 메소드를 새로 만들어내게 됩니다. 즉 인스턴스가 늘어날수록 성능에 부하를 일으키기 때문에 prototype inheritance를 이용하기로 했습니다.
